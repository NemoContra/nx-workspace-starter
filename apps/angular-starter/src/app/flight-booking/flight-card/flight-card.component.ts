import { Component, EventEmitter, Input, Output } from '@angular/core';
import type { Flight } from '@flight-app/shared';
import { RouterLink } from '@angular/router';
import { DatePipe, NgIf, NgStyle } from '@angular/common';

@Component({
  selector: 'app-flight-card',
  templateUrl: './flight-card.component.html',
  styleUrls: ['./flight-card.component.scss'],
  imports: [RouterLink, DatePipe, NgStyle, NgIf],
  standalone: true,
})
export class FlightCardComponent {
  @Input() item: Flight;
  @Input() selected: boolean;
  @Output() selectedChange = new EventEmitter<boolean>();

  select() {
    this.selected = true;
    this.selectedChange.next(this.selected);
  }

  deselect() {
    this.selected = false;
    this.selectedChange.next(this.selected);
  }
}
