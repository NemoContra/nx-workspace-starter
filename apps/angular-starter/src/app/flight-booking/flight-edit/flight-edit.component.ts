import { Component, inject, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';

import { FlightService } from '../flight-search/flight.service';
import type { Flight } from '@flight-app/shared';
import { FormsModule } from '@angular/forms';
import { NgIf } from '@angular/common';

@Component({
  selector: 'app-flight-edit',
  templateUrl: './flight-edit.component.html',
  styleUrls: ['./flight-edit.component.scss'],
  imports: [NgIf, FormsModule],
  standalone: true,
})
export class FlightEditComponent implements OnInit {
  id: string;
  showDetails: string;

  flight: Flight;
  errors: string;

  private route = inject(ActivatedRoute);
  private flightService = inject(FlightService);

  ngOnInit() {
    this.route.params.subscribe((params) => {
      this.id = params['id'];
      this.showDetails = params['showDetails'];

      this.flightService.findById(this.id).subscribe({
        next: (flight) => {
          this.flight = flight;
          this.errors = '';
        },
        error: () => {
          this.errors = 'Fehler beim Laden';
        },
      });
    });
  }

  save() {
    this.flightService.save(this.flight).subscribe({
      next: (flight) => {
        this.flight = flight;
        this.errors = 'Saving was successful!';
      },
      error: () => {
        this.errors = 'Error saving data';
      },
    });
  }
}
