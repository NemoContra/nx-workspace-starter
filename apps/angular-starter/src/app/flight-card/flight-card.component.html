<div [ngClass]="{ selected: selected }" class="card">
  <div class="card-header">
    <h2 class="title">
      {{ item.from | city : 'short' }} - {{ item.to | city : 'long' }}
    </h2>
  </div>

  <div class="card-body">
    <p>Flight-No.: #{{ item.id }}</p>
    <p>Date: {{ item.date | date : 'dd.MM.yyyy HH:mm' }}</p>
    <p [ngStyle]="{ color: item.delayed }">Delayed: {{ item.delayed }}</p>
    <p>
      <button
        (click)="select()"
        *ngIf="!selected"
        class="btn btn-default"
        data-testid="select-button"
      >
        Select
      </button>
      <button
        (click)="deselect()"
        *ngIf="selected"
        class="btn btn-default"
        data-testid="deselect-button"
      >
        Remove
      </button>
      <button
        class="btn btn-default"
        (click)="edit()"
        data-testid="edit-button"
      >
        Edit
      </button>
    </p>
  </div>
</div>
