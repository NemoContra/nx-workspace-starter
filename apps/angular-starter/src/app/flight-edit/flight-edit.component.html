<mat-dialog-content>
  <div class="p40">
    <h2>Flight Edit</h2>

    <form #form="ngForm" class="form-group">
      <div *ngIf="!form?.controls?.['from']?.valid" class="alert alert-danger">
        Validation errors in 'from'.
      </div>
      <div
        *ngIf="form?.controls?.['from']?.hasError('required')"
        class="alert alert-danger"
      >
        The field 'from' is required!
      </div>
      <div
        *ngIf="form?.controls?.['from']?.hasError('minlength')"
        class="alert alert-danger"
      >
        The field 'from' needs to have at least
        {{form?.controls?.['from']?.errors?.['minlength']?.requiredLength}}
        characters.
      </div>
      <div *ngIf="!form?.controls?.['from']?.valid" class="alert alert-danger">
        Internal error object: {{ form?.controls?.['from']?.errors | json }}
      </div>

      <div>
        <div class="row">
          <input
            placeholder="Id"
            class="form-control"
            name="id"
            [(ngModel)]="flight.id"
          />
        </div>
        <div class="row">
          <input
            placeholder="From"
            class="form-control"
            name="from"
            [(ngModel)]="flight.from"
            required
            minlength="3"
          />
        </div>
        <div class="row">
          <input
            placeholder="To"
            class="form-control"
            name="to"
            [(ngModel)]="flight.to"
          />
        </div>
        <div class="row">
          <input
            placeholder="Date"
            class="form-control"
            name="date"
            [(ngModel)]="flight.date"
          />
        </div>
      </div>

      <div class="row">
        <input
          name="delayed"
          class="form-control"
          id="delayed"
          [(ngModel)]="flight.delayed"
          type="checkbox"
        />
        <label for="delayed" class="form-label">Delayed</label>
      </div>

      <div>
        <button (click)="close()" class="btn btn-default">Close</button>
      </div>
    </form>
  </div>
</mat-dialog-content>
