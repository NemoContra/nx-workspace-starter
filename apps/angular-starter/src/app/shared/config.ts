export type Language = 'de' | 'en';

export interface Config {
  baseUrl: string;
  language: Language;
}

export const initConfig: Config = {
  baseUrl: '',
  language: 'de',
};
