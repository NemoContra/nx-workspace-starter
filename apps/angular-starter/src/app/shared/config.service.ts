import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Config, initConfig, Language } from './config';

@Injectable({
  providedIn: 'root',
})
export class ConfigService {
  private http = inject(HttpClient);
  private _config = initConfig;

  get config(): Config {
    return { ...this._config };
  }

  loadConfig() {
    this.http.get<Config>('./assets/config.json').subscribe((config) => {
      this._config = config;
    });
  }

  setLanguage(language: Language): void {
    this._config.language = language;
  }

  getLanguage(): Language {
    return this._config.language;
  }
}
