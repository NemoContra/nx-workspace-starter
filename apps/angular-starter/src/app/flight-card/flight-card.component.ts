import { Component, EventEmitter, inject, Input, Output } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Flight } from '@flight-app/shared';
import { CityPipe } from '../shared/city.pipe';
import { MatDialog } from '@angular/material/dialog';
import { FlightEditComponent } from '../flight-edit/flight-edit.component';

export const initFlight: Flight = {
  id: 0,
  from: '',
  to: '',
  date: '',
  delayed: false,
};

@Component({
  selector: 'flight-card',
  standalone: true,
  imports: [CommonModule, CityPipe],
  templateUrl: './flight-card.component.html',
  styleUrls: ['./flight-card.component.scss'],
})
export class FlightCardComponent {
  @Input() item: Flight = initFlight;
  @Input() selected = false;
  @Output() selectedChange = new EventEmitter<boolean>();

  private dialog = inject(MatDialog);

  select(): void {
    this.selected = true;
    this.selectedChange.emit(this.selected);
  }

  deselect(): void {
    this.selected = false;
    this.selectedChange.emit(this.selected);
  }

  edit() {
    this.dialog.open(FlightEditComponent, {
      data: { flight: this.item },
      minWidth: '80%',
    });
  }
}
