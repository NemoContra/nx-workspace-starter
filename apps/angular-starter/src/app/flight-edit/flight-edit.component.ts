import { Component, inject } from '@angular/core';
import { CommonModule } from '@angular/common';
import {
  MAT_DIALOG_DATA,
  MatDialogModule,
  MatDialogRef,
} from '@angular/material/dialog';
import { Flight } from '@flight-app/shared';
import { FormsModule } from '@angular/forms';
import { MatCardModule } from '@angular/material/card';

@Component({
  selector: 'flight-edit',
  standalone: true,
  imports: [CommonModule, FormsModule, MatCardModule, MatDialogModule],
  templateUrl: './flight-edit.component.html',
  styleUrls: ['./flight-edit.component.scss'],
})
export class FlightEditComponent {
  private dialogRef = inject(MatDialogRef);
  private data = inject<{ flight: Flight }>(MAT_DIALOG_DATA);

  flight = this.data.flight;

  close(): void {
    this.dialogRef.close();
  }
}
