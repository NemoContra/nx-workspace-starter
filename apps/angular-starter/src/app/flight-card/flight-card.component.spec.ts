import { FlightCardComponent } from './flight-card.component';
import { MatDialog } from '@angular/material/dialog';
import {
  byTestId,
  createComponentFactory,
  Spectator,
} from '@ngneat/spectator/jest';
import { Flight } from '@flight-app/shared';
import { FlightEditComponent } from '../flight-edit/flight-edit.component';

const mockFlight: Flight = {
  from: 'Hamburg',
  to: 'Rankweil',
  date: '2023-07-20',
  delayed: false,
  id: 1,
};

describe('FlightCardComponent', () => {
  const createComponent = createComponentFactory({
    component: FlightCardComponent,
    providers: [MatDialog],
  });
  let spectator: Spectator<FlightCardComponent>;

  beforeEach(() => {
    spectator = createComponent({
      props: {
        item: mockFlight,
      },
    });
  });

  it('should render', () => {
    expect(spectator.fixture).toMatchSnapshot();
  });

  it('should open a dialog if edit button is clicked', () => {
    const dialog = spectator.inject(MatDialog);
    jest.spyOn(dialog, 'open');

    expect(spectator.query(byTestId('edit-button'))).toBeTruthy();
    spectator.click(byTestId('edit-button'));

    expect(dialog.open).toHaveBeenCalledTimes(1);
    expect(dialog.open).toHaveBeenCalledWith(FlightEditComponent, {
      data: {
        flight: mockFlight,
      },
      minWidth: '80%',
    });
  });

  it('should select a flight', () => {
    const spyObserver = {
      next: jest.fn(),
      error: jest.fn(),
    };

    spectator.output('selectedChange').subscribe(spyObserver);

    spectator.click(byTestId('select-button'));
    expect(spectator.fixture).toMatchSnapshot();

    expect(spyObserver.next).toHaveBeenCalledTimes(1);
    expect(spyObserver.next).toHaveBeenCalledWith(true);
  });

  it('should deselect a flight', () => {
    const spyObserver = {
      next: jest.fn(),
      error: jest.fn(),
    };

    spectator.output('selectedChange').subscribe(spyObserver);

    spectator.click(byTestId('select-button'));
    spectator.click(byTestId('deselect-button'));

    expect(spyObserver.next).toHaveBeenCalledTimes(2);
    expect(spyObserver.next).toHaveBeenNthCalledWith(1, true);
    expect(spyObserver.next).toHaveBeenNthCalledWith(2, false);
  });
});
