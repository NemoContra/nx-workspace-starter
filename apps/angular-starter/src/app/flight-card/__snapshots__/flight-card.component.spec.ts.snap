// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightCardComponent should render 1`] = `
<flight-card
  dialog={[Function Object]}
  item={[Function Object]}
  selected="false"
  selectedChange={[Function EventEmitter_]}
>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <h2
        class="title"
      >
         Hamburg - Rankweil 
      </h2>
    </div>
    <div
      class="card-body"
    >
      <p>
        Flight-No.: #1
      </p>
      <p>
        Date: 20.07.2023 00:00
      </p>
      <p>
        Delayed: false
      </p>
      <p>
        <button
          class="btn btn-default"
          data-testid="select-button"
        >
           Select 
        </button>
        <button
          class="btn btn-default"
          data-testid="edit-button"
        >
           Edit 
        </button>
      </p>
    </div>
  </div>
</flight-card>
`;

exports[`FlightCardComponent should select a flight 1`] = `
<flight-card
  dialog={[Function Object]}
  item={[Function Object]}
  selected={[Function Boolean]}
  selectedChange={[Function EventEmitter_]}
>
  <div
    class="card selected"
  >
    <div
      class="card-header"
    >
      <h2
        class="title"
      >
         Hamburg - Rankweil 
      </h2>
    </div>
    <div
      class="card-body"
    >
      <p>
        Flight-No.: #1
      </p>
      <p>
        Date: 20.07.2023 00:00
      </p>
      <p>
        Delayed: false
      </p>
      <p>
        <button
          class="btn btn-default"
          data-testid="deselect-button"
        >
           Remove 
        </button>
        <button
          class="btn btn-default"
          data-testid="edit-button"
        >
           Edit 
        </button>
      </p>
    </div>
  </div>
</flight-card>
`;
